File: 048_scatter_plot_qtcb/R/build_data.R
# build_data.R -- Example 048: QTcB
set.seed(48); N_SUBJ<-60; TRTS<-c("Placebo","Drug A 10mg","Drug A 25mg"); VISITS<-1:8
trt_traj<-list(Placebo=c(400,401,402,402,403,403,402,401),`Drug A 10mg`=c(400,405,410,415,418,420,418,415),`Drug A 25mg`=c(400,408,418,428,435,440,438,432))
records<-list(); subj_n<-1L
for(trt in TRTS){ means<-trt_traj[[trt]]
  for(i in seq_len(N_SUBJ)){ subj_re<-rnorm(1,0,15)
    for(vi in seq_along(VISITS)){ v<-VISITS[vi]; if(v>4&&runif(1)<.08) next
      records[[length(records)+1]]<-data.frame(subjid=sprintf("S%04d",subj_n),trt=trt,visit=v,qtcb=round(means[vi]+subj_re+rnorm(1,0,12),1),stringsAsFactors=FALSE); subj_n<-subj_n+1L}}}
df<-do.call(rbind,records); write.csv(df,"test_data.csv",row.names=FALSE); message(sprintf("Generated %d rows.",nrow(df)))

File: 048_scatter_plot_qtcb/python/build_data.py
"""build_data.py -- Example 048: QTcB over time by treatment."""
import numpy as np, pandas as pd
rng=np.random.default_rng(seed=48); N_SUBJ=60; TRTS=["Placebo","Drug A 10mg","Drug A 25mg"]; VISITS=list(range(1,9))
trt_traj={"Placebo":[400,401,402,402,403,403,402,401],"Drug A 10mg":[400,405,410,415,418,420,418,415],"Drug A 25mg":[400,408,418,428,435,440,438,432]}
records=[]; subj=1
for trt in TRTS:
    means=trt_traj[trt]
    for _ in range(N_SUBJ):
        subj_re=rng.normal(0,15)
        for vi,v in enumerate(VISITS):
            if v>4 and rng.random()<.08: continue
            qtcb=means[vi]+subj_re+rng.normal(0,12); records.append({"subjid":f"S{subj:04d}","trt":trt,"visit":v,"qtcb":round(qtcb,1)}); subj+=1
df=pd.DataFrame(records); df.to_csv("test_data.csv",index=False); print(f"Generated {len(df)} rows.")
